test/integration:  Add a test for the asyncprofiler package

v-thakkar opened the pull request:
This PR adds an integration test for the asyncprofiler package as part of https://github.com/parca-dev/parca-agent/issues/1510. The PR also does the necessary adjustments in the github workflow files to make the test work. The test flow looks like the following:

- Starts the java app
- Loads the fdtransfer binary
- Attaches and loads libasyncprofiler.so to start the profiling
- Collects the output in .jfr
- Stops the profiling after certain duration
- Converts jfr file to pprof format
- Cleans up the files and processes

Test Plan:
-------------
Parca-agent CI

v-thakkar commented on .pre-commit-config.yaml:
```diff
@@ -5,7 +5,7 @@ repos:
         exclude: ^internal/(?:go|pprof)/
       - id: trailing-whitespace
-        exclude: ^internal/(?:go|pprof)/
+        exclude: ^internal/(?:go|pprof)/|^tests/testdata/
```
@maxbrunet Hi :wave: Any idea why the pre-commit-lite bot is not taking the changes of .pre-commit-config into the account? I thought these changes combined with the ones in .golangci should make the pre-commit-lite bot from making any changes in the PR. What am I doing wrong?

v-thakkar replied:
I had also tried `paths_ignore` from here yesterday: https://github.com/marketplace/actions/skip-duplicate-actions#skip-ignored-paths but that didn't seem to work either :(

maxbrunet commented on .pre-commit-config.yaml:
```diff
@@ -5,7 +5,7 @@ repos:
       - id: end-of-file-fixer
         exclude: ^internal/(?:go|pprof)/
```
@v-thakkar This one is making the change

```
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing tests/testdata/java-app/src/main/resources/application.properties
```

We could move to a single `exclude` at the top-level:

https://github.com/parca-dev/demo-deployments/blob/667f077ca351965812a736a6415192ef5572ee96/.pre-commit-config.yaml#L1

v-thakkar replied:
Aha, ok. Let me try doing that. And see if it works out. Thanks for the pointer!

v-thakkar replied:
Yay, that worked. Thanks for unblocking me on that front. I'll take a look at why go build fails a little later in the evening.

v-thakkar commented:
On a host:

```
vaishali@pop-os:~/parca-agent/tests$ java --version
openjdk 17.0.7 2023-04-18 LTS
OpenJDK Runtime Environment Corretto-17.0.7.7.1 (build 17.0.7+7-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.7.7.1 (build 17.0.7+7-LTS, mixed mode, sharing)
```
```
vaishali@pop-os:~/parca-agent/tests$ go test
Executing command: /usr/bin/java -Xms1G -Xmx1G -XX:+AlwaysPreTouch -jar /home/vaishali/parca-agent/tests/testdata/java-app/target/first-application-0.0.1-SNAPSHOT.jar
Java PID: 18139
Executing command: /home/vaishali/parca-agent/tests/testdata/jattach 18139 load /home/vaishali/parca-agent/tests/testdata/libasyncProfiler.so true start,event=cpu,file=prof.jfr
Executing command: /home/vaishali/parca-agent/tests/testdata/jattach 18139 load /home/vaishali/parca-agent/tests/testdata/libasyncProfiler.so true stop
PASS
ok  	github.com/parca-dev/parca-agent/tests	180.271s
```

v-thakkar commented:
Strace log from the gh-actions after setting LD_LIBRARY_PATH to default standard values.
[strace_log_ghaction.txt](https://github.com/parca-dev/parca-agent/files/11476949/strace_log_ghaction.txt)

v-thakkar commented on .github/workflows/build.yml:
```diff
@@ -100,6 +100,10 @@ jobs:
         run: unset LD_LIBRARY_PATH
 
+      - name: Allow perf_events to be accessed by non-root users
+        # TODO(vthakkar): Fix the bug in FDTransfer to handle this
```
JFYI, will handle this todo in the follow-up PR. Also, I'm keeping the debugging logs in the test for now as they're useful.

brancz commented on test/integration/testdata/fdtransfer:
the idea was to use the download script for this so we don't have to check in binaries

brancz commented on test/integration/testdata/jattach:
same as fdtransfer

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	)
+
+	// Run fdtransfer
+	err = asyncprofiler.RunFdtransfer(filepath.Join(cwd, "testdata", "fdtransfer"), javaPID)
```
I thought we do the `sysctl` so we don't need to do this?

v-thakkar replied:
sysctl is only used for now. We do want this to work, right? I wanted to still keep this so that when fdtransfer related bug is fixed, all we need to do is remove sysctl from the github action. But if you want me to add this only when it works, I can remove it.

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	err = cmd.Start()
+	require.NoError(t, err, "Failed to start Java program")
+
+	time.Sleep(50 * time.Second) // We need to wait for the Java program to start
```
What is it that we're actually waiting for here? How would we tell from the "production" profiler that it's safe to attach?

v-thakkar replied:
Oops, sorry. These are leftovers. They were just part of one of the debugging. Same for the ones below. Will remove them.

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	err = asyncprofiler.RunFdtransfer(filepath.Join(cwd, "testdata", "fdtransfer"), javaPID)
+	require.NoError(t, err, "Failed to run fdtransfer")
+
+	time.Sleep(10 * time.Second) // Wait for fdtransfer to finish loading
```
What does "finish loading" mean? Is there something specific that we could watch for in a shorter interval?

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	require.NoError(t, err, "Failed to run start command")
+
+	// Sleep for the specified duration
+	time.Sleep(120 * time.Second)
```
same as before, I think we're waiting for the jfr file to appear?

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	outputFiles := []string{jfrOutputFile, "prof.pprof"}
+
+	for _, outputFile := range outputFiles {
+		err = os.Remove(outputFile)
```
the typical thing to do here is a `defer os.Remove(file)` right after creating the file just like it's already done with the `.Close()`.

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+
+	// Terminate the Java program
+	process := cmd.Process
+	err = process.Signal(syscall.SIGKILL)
```
this should probably be done in a defer as well

brancz commented on test/integration/asyncprofiler_test.go:
```diff
@@ -0,0 +1,138 @@
+	require.NotNil(t, pprofProfile, "Pprof profile is nil")
+
+	// Save the converted pprof profile to a file
+	pprofFile, err := os.Create("prof.pprof")
```
I don't think we need to do this, we can stop at the point where JRF to pprof conversion doesn't fail. However, we should check that the profile created is not empty (as in the `.Sample` is non 0 length).

brancz commented on test/integration/testdata/java-app/prof.jfr:
I think this was accidentally checked in?

v-thakkar replied:
Yes :see_no_evil: Good catch!

brancz reviewed (commented):
Going in the right direction. I wonder if we really need a java example that's this complex, what's wrong with the java example from parca-demo?

We should also remove all the compiled artifacts from being checked in.

v-thakkar commented:
For the Java example, I added this one for two resoans:
1. So we don't have to keep changing it as we exercise further options, like memory or wall clock profiling.
2. Another reason for this specific iteration was not to spend much time on making the Java app work for Corretto 17. So I picked the one which I knew already worked and has already been tested with Corretto 17. I did want to add the smaller binary in the demo for Corretto 17 eventually once Fdtransfer works. But I can prioritize it first.

v-thakkar commented:
Things seem to have changed in the CI, checking what is the reason for the new failure. Local testinmg;

```
vaishali@pop-os:~/parca-agent$ sudo make test/integration GO=$(which go) ARCH=amd64
mkdir -p pkg/profiler/cpu/bpf/amd64
make -C bpf build
make[1]: Entering directory '/home/vaishali/parca-agent/bpf'
make[1]: Nothing to be done for 'build'.
make[1]: Leaving directory '/home/vaishali/parca-agent/bpf'
cp bpf/out/amd64/cpu.bpf.o pkg/profiler/cpu/bpf/amd64/cpu.bpf.o
sudo CGO_ENABLED=1 GOOS=linux GOARCH=amd64 CC="clang" CGO_CFLAGS="-I/home/vaishali/parca-agent/dist/libbpf/amd64/usr/include" CGO_LDFLAGS="-fuse-ld=ld -lzstd /home/vaishali/parca-agent/dist/libbpf/amd64/libbpf.a" /usr/local/go/bin/go test  -v ./test/integration/... -count=1
=== RUN   TestIntegrationAsyncProfiler
Executing command: /usr/bin/java -Xms1G -Xmx1G -XX:+AlwaysPreTouch -jar /home/vaishali/parca-agent/test/integration/testdata/java-app/target/first-application-0.0.1-SNAPSHOT.jar
Java PID: 2711033
Executing command: /home/vaishali/parca-agent/test/integration/testdata/jattach 2711033 load /home/vaishali/parca-agent/test/integration/testdata/libasyncProfiler.so true start,event=cpu,file=/home/vaishali/parca-agent/test/integration/prof.jfr
Executing command: /home/vaishali/parca-agent/test/integration/testdata/jattach 2711033 load /home/vaishali/parca-agent/test/integration/testdata/libasyncProfiler.so true stop
--- PASS: TestIntegrationAsyncProfiler (180.35s)
=== RUN   TestAnyStackContains
--- PASS: TestAnyStackContains (0.00s)
=== RUN   TestCPUProfilerWorks
--- PASS: TestCPUProfilerWorks (22.70s)
PASS
ok  	github.com/parca-dev/parca-agent/test/integration	203.095s
```

v-thakkar commented:
Test passes on Sumera's machine as well:
![image_2](https://github.com/parca-dev/parca-agent/assets/7105009/91bdd6d3-a3e0-4652-b8a0-70b9436856f3)
Will try tmate to see what is it this time!

v-thakkar commented:
Ok, I think this is ready for the next round of reviews. I've replaced Java app with the one we're using in the demo. I've added compiled JAR here along with the app but if we're fine with adding it in the demo repo, I can move it there.

I'll handle the TODOs in a follow-up PR(s). I keep hitting [this issue](https://github.com/parca-dev/parca-agent/issues/1798) while running integration tests locally which is quite annoying so would like to probably handle that first before handling TODOs.
cc @brancz

joeyleeeeeee97 commented:
@brancz @v-thakkar 

Hi , is there any updates on this? I am looking forward to use this function

kakkoyun commented:
> @brancz @v-thakkar
> 
> Hi , is there any updates on this? I am looking forward to use this function

We plan to focus on the Java support in this quarter (2024 Q1). We will update this PR accordingly.

kakkoyun commented:
We changed direction for Java profiling. So I'm closing this one.
